proc/orderedmerge.go

```diff
@@ -0,0 +1,89 @@
+package proc
+
+import (
+	"sync"
+
+	"github.com/brimsec/zq/zbuf"
+)
+
+// An OrderedMerge proc merges multiple upstream inputs into one output.
+// If the input streams are ordered according to the configured comparison,
+// the output of OrderedMerge will have the same order.
+type OrderedMerge struct {
+	Base
+	batches   []zbuf.Batch
+	compare   zbuf.BatchCmpFn
+	done      []bool
+	once      sync.Once
+	parents   []Proc
+	resultChs []chan Result
+}
+
+func NewOrderedMerge(c *Context, parents []Proc, cmp zbuf.BatchCmpFn) *OrderedMerge {
+	n := len(parents)
+	m := &OrderedMerge{
+		Base:      Base{Context: c},
+		batches:   make([]zbuf.Batch, n),
+		compare:   cmp,
+		done:      make([]bool, n),
+		parents:   parents,
+		resultChs: make([]chan Result, n),
+	}
+	for i := range m.resultChs {
+		m.resultChs[i] = make(chan Result)
+	}
+	return m
+}
+
+func (m *OrderedMerge) run() {
+	for i := range m.parents {
+		i := i
+		go func() {
+			for {
+				batch, err := m.parents[i].Pull()
+				select {
+				case m.resultChs[i] <- Result{batch, err}:
+					if EOS(batch, err) {
+						return
+					}
+				case <-m.Context.Done():
+					return
+				}
+			}
+		}()
+	}
+}
+
+func (m *OrderedMerge) Pull() (zbuf.Batch, error) {
+	m.once.Do(m.run)
+	idx := -1
+	for i := range m.batches {
+		if m.done[i] {
+			continue
+		}
+		if m.batches[i] == nil {
+			select {
+			case res := <-m.resultChs[i]:
+				if res.Err != nil {
+					return nil, res.Err
+				}
+				if res.Batch == nil {
+					m.done[i] = true
+					continue
+				}
+				m.batches[i] = res.Batch
+			case <-m.Context.Done():
+				return nil, m.Context.Err()
+			}
+		}
+		if idx == -1 || m.compare(m.batches[i], m.batches[idx]) {
```

Comment:
Unless I'm missing something, I don't think we can always use batch-by-batch comparison here. I think we need to do record by record (like `zbuf.Combiner` does, and like sort/groupby do when reading from spills). 

(Batch-by-batch could be an optimization for cases where we know that batches are non-overlapping in the ordering key, but I'd suggest we add that at a later stage if necessary).

Reply:
Thanks, I'll adjust. For the use case with archives, I was working on the assumption that there is no overlap between the batches we use for the parallel scanner, but that's a very narrow use case that doesn't work well in this general proc.

Reply:
I've added the per-record changes & a test, PTAL!